Orphan parameter reference in capture set

## Compiler version

`cc-experiment`

## Minimized code

<!--
This code should be self contained, compilable (with possible failures) and as small as possible.

Ideally, we should be able to just copy this code in a file and run `scalac` (and maybe `scala`) to reproduce the issue.
-->

```Scala
trait Cap { def use(): Int }
type Id[X] = [T] -> (op: X => T) -> T
def mkId[X](x: X): Id[X] = [T] => (op: X => T) => op(x)

def withCap[X](op: ({*} Cap) => X): X = {
  val cap: {*} Cap = new Cap { def use() = { println("cap is used"); 0 } }
  val result = op(cap)
  result
}

def leaking(c: {*} Cap): Id[{c} Cap] = mkId(c)

val bad = withCap(leaking)
```

## Output

<details>

```scala
error when pickling type c.type
error when pickling tree c
error when pickling tree [c : Any]
error when pickling tree [c : Any]*
error when pickling tree new _root_.scala.annotation.retains([c : Any]*)
error when pickling type {c} Cap
error when pickling type Id[{c} Cap]
error when pickling tree Id[{c} Cap]
error when pickling tree val bad: Id[{c} Cap] = 
  withCap[Id[{c} Cap]](
    {
      def $anonfun(c: {*} Cap): Id[{c} Cap] = leaking(c)
      closure($anonfun)
    }
  )
error when pickling tree () extends Object() { this: orphan-parameref$package.type =>
  private def writeReplace(): AnyRef = new scala.runtime.ModuleSerializationProxy(classOf[orphan-parameref$package.type])
  type Id[X >: Nothing <: Any] = 
    PolyFunction 
      {
        def apply[T >: Nothing <: Any](op: X => T): T
      }
  def mkId[X >: Nothing <: Any](x: X): Id[X] = 
    {
      final class $anon() extends Object(), PolyFunction {
        def apply[T >: Nothing <: Any](op: X => T): T = op.apply(x)
      }
      new $anon():([T] => (op: X => T) -> T)
    }
  def withCap[X >: Nothing <: Any](op: ({*} Cap) => X): X = 
    {
      val cap: {*} Cap = 
        {
          final class $anon() extends Object(), Cap {
            def use(): Int = 
              {
                println("cap is used")
                0
              }
          }
          new $anon():Cap
        }
      val result: X = op.apply(cap)
      result:X
    }
  def leaking(c: {*} Cap): Id[{c} Cap] = mkId[({*} Cap)](c)
  val bad: Id[{c} Cap] = 
    withCap[Id[{c} Cap]](
      {
        def $anonfun(c: {*} Cap): Id[{c} Cap] = leaking(c)
        closure($anonfun)
      }
    )
}
error when pickling tree @CaptureChecked @SourceFile("issues/orphan-parameref.scala") final module class orphan-parameref$package() extends Object() { this: orphan-parameref$package.type =>
  private def writeReplace(): AnyRef = new scala.runtime.ModuleSerializationProxy(classOf[orphan-parameref$package.type])
  type Id[X >: Nothing <: Any] = 
    PolyFunction 
      {
        def apply[T >: Nothing <: Any](op: X => T): T
      }
  def mkId[X >: Nothing <: Any](x: X): Id[X] = 
    {
      final class $anon() extends Object(), PolyFunction {
        def apply[T >: Nothing <: Any](op: X => T): T = op.apply(x)
      }
      new $anon():([T] => (op: X => T) -> T)
    }
  def withCap[X >: Nothing <: Any](op: ({*} Cap) => X): X = 
    {
      val cap: {*} Cap = 
        {
          final class $anon() extends Object(), Cap {
            def use(): Int = 
              {
                println("cap is used")
                0
              }
          }
          new $anon():Cap
        }
      val result: X = op.apply(cap)
      result:X
    }
  def leaking(c: {*} Cap): Id[{c} Cap] = mkId[({*} Cap)](c)
  val bad: Id[{c} Cap] = 
    withCap[Id[{c} Cap]](
      {
        def $anonfun(c: {*} Cap): Id[{c} Cap] = leaking(c)
        closure($anonfun)
      }
    )
}
error when pickling tree package <empty> {
  final lazy module val orphan-parameref$package: orphan-parameref$package = new orphan-parameref$package()
  @CaptureChecked @SourceFile("issues/orphan-parameref.scala") final module class orphan-parameref$package() extends Object() { this: orphan-parameref$package.type =>
    private def writeReplace(): AnyRef = new scala.runtime.ModuleSerializationProxy(classOf[orphan-parameref$package.type])
    type Id[X >: Nothing <: Any] = 
      PolyFunction 
        {
          def apply[T >: Nothing <: Any](op: X => T): T
        }
    def mkId[X >: Nothing <: Any](x: X): Id[X] = 
      {
        final class $anon() extends Object(), PolyFunction {
          def apply[T >: Nothing <: Any](op: X => T): T = op.apply(x)
        }
        new $anon():([T] => (op: X => T) -> T)
      }
    def withCap[X >: Nothing <: Any](op: ({*} Cap) => X): X = 
      {
        val cap: {*} Cap = 
          {
            final class $anon() extends Object(), Cap {
              def use(): Int = 
                {
                  println("cap is used")
                  0
                }
            }
            new $anon():Cap
          }
        val result: X = op.apply(cap)
        result:X
      }
    def leaking(c: {*} Cap): Id[{c} Cap] = mkId[({*} Cap)](c)
    val bad: Id[{c} Cap] = 
      withCap[Id[{c} Cap]](
        {
          def $anonfun(c: {*} Cap): Id[{c} Cap] = leaking(c)
          closure($anonfun)
        }
      )
  }
}
exception occurred while compiling issues/orphan-parameref.scala
java.lang.AssertionError: assertion failed: orphan parameter reference: TermParamRef(c) while compiling issues/orphan-parameref.scala
Exception in thread "main" java.lang.AssertionError: assertion failed: orphan parameter reference: TermParamRef(c)
	at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleNewType(TreePickler.scala:291)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleType(TreePickler.scala:160)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:384)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$17$$anonfun$1(TreePickler.scala:513)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$17(TreePickler.scala:513)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$adapted$16(TreePickler.scala:513)
	at scala.Function0.apply$mcV$sp(Function0.scala:39)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:58)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:513)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$6(TreePickler.scala:463)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$adapted$6(TreePickler.scala:463)
	at scala.Function0.apply$mcV$sp(Function0.scala:39)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:58)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:463)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$3$$anonfun$1(TreePickler.scala:433)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$3(TreePickler.scala:433)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$adapted$3(TreePickler.scala:434)
	at scala.Function0.apply$mcV$sp(Function0.scala:39)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:58)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:434)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleNewType$$anonfun$6(TreePickler.scala:263)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleNewType$$anonfun$adapted$6(TreePickler.scala:263)
	at scala.Function0.apply$mcV$sp(Function0.scala:39)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:58)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleNewType(TreePickler.scala:263)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleType(TreePickler.scala:160)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleNewType$$anonfun$2$$anonfun$1(TreePickler.scala:181)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleNewType$$anonfun$2(TreePickler.scala:181)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleNewType$$anonfun$adapted$2(TreePickler.scala:181)
	at scala.Function0.apply$mcV$sp(Function0.scala:39)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:58)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleNewType(TreePickler.scala:181)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleType(TreePickler.scala:160)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:609)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTpt(TreePickler.scala:318)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleDef$$anonfun$1(TreePickler.scala:336)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleDef$$anonfun$adapted$1(TreePickler.scala:340)
	at scala.Function0.apply$mcV$sp(Function0.scala:39)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:58)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleDef(TreePickler.scala:340)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:550)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleStats$$anonfun$2(TreePickler.scala:365)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleStats(TreePickler.scala:365)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$26(TreePickler.scala:591)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$adapted$24(TreePickler.scala:592)
	at scala.Function0.apply$mcV$sp(Function0.scala:39)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:58)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:592)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleDef$$anonfun$1(TreePickler.scala:335)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleDef$$anonfun$adapted$1(TreePickler.scala:340)
	at scala.Function0.apply$mcV$sp(Function0.scala:39)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:58)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleDef(TreePickler.scala:340)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:567)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleStats$$anonfun$2(TreePickler.scala:365)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleStats(TreePickler.scala:365)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$29(TreePickler.scala:607)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$adapted$27(TreePickler.scala:607)
	at scala.Function0.apply$mcV$sp(Function0.scala:39)
	at dotty.tools.dotc.core.tasty.TreePickler.withLength(TreePickler.scala:58)
	at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:607)
	at dotty.tools.dotc.core.tasty.TreePickler.pickle$$anonfun$1(TreePickler.scala:779)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.dotc.core.tasty.TreePickler.pickle(TreePickler.scala:779)
	at dotty.tools.dotc.transform.Pickler.run$$anonfun$1$$anonfun$1(Pickler.scala:72)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.dotc.transform.Pickler.run$$anonfun$1(Pickler.scala:110)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at dotty.tools.dotc.transform.Pickler.run(Pickler.scala:110)
	at dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:316)
	at scala.collection.immutable.List.map(List.scala:246)
	at dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:317)
	at dotty.tools.dotc.transform.Pickler.runOn(Pickler.scala:115)
	at dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:233)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
	at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
	at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1328)
	at dotty.tools.dotc.Run.runPhases$1(Run.scala:244)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:252)
	at dotty.tools.dotc.Run.compileUnits$$anonfun$adapted$1(Run.scala:261)
	at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:68)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:261)
	at dotty.tools.dotc.Run.compileSources(Run.scala:185)
	at dotty.tools.dotc.Run.compile(Run.scala:169)
	at dotty.tools.dotc.Driver.doCompile(Driver.scala:35)
	at dotty.tools.dotc.Driver.process(Driver.scala:195)
	at dotty.tools.dotc.Driver.process(Driver.scala:163)
	at dotty.tools.dotc.Driver.process(Driver.scala:175)
	at dotty.tools.dotc.Driver.main(Driver.scala:205)
	at dotty.tools.dotc.Main.main(Main.scala)
```

</details>

## Expectation

It should not trigger `AssertionError`.

--- a/compiler/src/dotty/tools/dotc/ast/TreeTypeMap.scala
+++ b/compiler/src/dotty/tools/dotc/ast/TreeTypeMap.scala
@@ -56,7 +56,7 @@ class TreeTypeMap(
   /** Replace occurrences of `This(oldOwner)` in some prefix of a type
    *  by the corresponding `This(newOwner)`.
    */
-  private val mapOwnerThis = new TypeMap {
+  private val mapOwnerThis = new TypeMap with cc.CaptureSet.IdempotentCaptRefMap {
     private def mapPrefix(from: List[Symbol], to: List[Symbol], tp: Type): Type = from match {
       case Nil => tp
       case (cls: ClassSymbol) :: from1 => mapPrefix(from1, to.tail, tp.substThis(cls, to.head.thisType))

--- a/compiler/src/dotty/tools/dotc/cc/CaptureSet.scala
+++ b/compiler/src/dotty/tools/dotc/cc/CaptureSet.scala
@@ -523,7 +523,7 @@ object CaptureSet:
 
     private def mapIsIdempotent = tm.isInstanceOf[IdempotentCaptRefMap]
 
-    assert(ccAllowUnsoundMaps || mapIsIdempotent)
+    assert(ccAllowUnsoundMaps || mapIsIdempotent, tm.getClass)
 
     private def whereCreated(using Context): String =
       if stack == null then ""

--- a/compiler/src/dotty/tools/dotc/core/Annotations.scala
+++ b/compiler/src/dotty/tools/dotc/core/Annotations.scala
@@ -1,4 +1,5 @@
-package dotty.tools.dotc
+package dotty.tools
+package dotc
 package core
 
 import Symbols._, Types._, Contexts._, Constants._
@@ -58,7 +59,7 @@ object Annotations {
             if tm.isRange(x) then x
             else
               val tp1 = tm(tree.tpe)
-              foldOver(if tp1 =:= tree.tpe then x else tp1, tree)
+              foldOver(if tp1 frozen_=:= tree.tpe then x else tp1, tree)
         val diff = findDiff(NoType, args)
         if tm.isRange(diff) then EmptyAnnotation
         else if diff.exists then derivedAnnotation(tm.mapOver(tree))
@@ -69,7 +70,7 @@ object Annotations {
       val args = arguments
       if args.isEmpty then false
       else tree.existsSubTree {
-        case id: Ident => id.tpe match
+        case id: Ident => id.tpe.stripped match
           case TermParamRef(tl1, _) => tl eq tl1
           case _ => false
         case _ => false

--- a/compiler/test/dotc/pos-test-pickling.blacklist
+++ b/compiler/test/dotc/pos-test-pickling.blacklist
@@ -19,6 +19,7 @@ i6507b.scala
 i12299a.scala
 i13871.scala
 i15181.scala
+i15922.scala
 
 # Tree is huge and blows stack for printing Text
 i7034.scala

--- a/tests/pos-custom-args/captures/i15922.scala
+++ b/tests/pos-custom-args/captures/i15922.scala
@@ -1,0 +1,14 @@
+trait Cap { def use(): Int }
+type Id[X] = [T] -> (op: X => T) -> T
+def mkId[X](x: X): Id[X] = [T] => (op: X => T) => op(x)
+
+def withCap[X](op: ({*} Cap) => X): X = {
+  val cap: {*} Cap = new Cap { def use() = { println("cap is used"); 0 } }
+  val result = op(cap)
+  result
+}
+
+def leaking(c: {*} Cap): Id[{c} Cap] = mkId(c)
+
+def test =
+  val bad = withCap(leaking)

--- a/tests/pos/i15922.scala
+++ b/tests/pos/i15922.scala
@@ -1,0 +1,14 @@
+trait Cap:
+  type M
+class Id[X]
+
+object Test:
+  def withCap[X](op: Cap => X): X = ???
+
+  class retains1(xs: Any*) extends annotation.StaticAnnotation
+
+  def leaking1(c: Cap): Id[Cap @retains1(c)] = ??? // used to crash with orphan parameter on pickling
+  def leaking2(c: Cap): Id[c.type] = ???
+
+  val bad1 = withCap(leaking1)
+  val bad2 = withCap(leaking2)

